pkg/sbom/package.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2022 Chainguard, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package sbom captures the internal data model of the SBOMs melange produces
+// into a private, generalized bill of materials model (with relationship data)
+// designed to be converted to specific formats — for now, just SPDX.
+package sbom
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+
+	"chainguard.dev/apko/pkg/sbom/generator/spdx"
+	"github.com/chainguard-dev/clog"
+	"github.com/github/go-spdx/v2/spdxexp"
+	purl "github.com/package-url/packageurl-go"
+	"golang.org/x/text/cases"
+	"golang.org/x/text/language"
+)
+
+// Package is a representation of an SBOM package specified by the build
+// process. It is later converted to an SPDX package, but it doesn't expose
+// fields that are invariant in the SPDX output.
+type Package struct {
+	// The name of the origin package, a subpackage, or any other kind of (e.g.
+	// non-APK) package for inclusion in the SBOM.
+	Name string
+
+	// The version of the package. For APK packages, this should be the "full
+	// version" (including the epoch).
+	Version string
+
+	// This is the copyright text in the SPDX package. It's usually left blank.
+	Copyright string
+
+	// SPDX license expression. Leaving this empty will result in NOASSERTION being
+	// used as its value.
+	LicenseDeclared string
+
+	// Name of the distro/organization that produced the package. E.g. "wolfi".
+	//
+	// TODO: consider renaming this to avoid confusion from our other uses of
+	//  "namespace", perhaps to "supplier" or "originator" (or have both), and signal
+	//  that it's safe to leave this blank.
+	Namespace string
+
+	// The architecture of the package. E.g. "aarch64". This field isn't always
+	// relevant, especially when describing material upstream of the built APK
+	// package (e.g. source code or language ecosystem dependencies).
+	Arch string
+
+	// Checksums of the package. The keys are the checksum algorithms (e.g. "SHA-256"),
+	// and the values are the checksums.
+	//
+	// TODO: We're not currently using this field, consider removing it.
+	Checksums map[string]string
```

Comment:
SPDX says do not include checksum for package that includes the SPDX itself. So we can start using this, if we move sbom from data tarball to control tarball and set this to like data tarball sha256sum.

Reply:
Sounds good, we can remove it for now and add it back when we know what we end up doing. This part of the code is a little awkward too, so it's probably good to approach it fresh with a clear use case.